sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala

```diff
@@ -116,13 +116,26 @@ object FileFormatWriter extends Logging {
     val partitionSet = AttributeSet(partitionColumns)
     val dataColumns = allColumns.filterNot(partitionSet.contains)
 
-    val bucketIdExpression = bucketSpec.map { spec =>
+    val bucketPartitioning = bucketSpec.map { spec =>
       val bucketColumns = spec.bucketColumnNames.map(c => dataColumns.find(_.name == c).get)
+      HashPartitioning(bucketColumns, spec.numBuckets)
+    }
+
+    val bucketIdExpression = bucketPartitioning.map { partitioning =>
       // Use `HashPartitioning.partitionIdExpression` as our bucket id expression, so that we can
       // guarantee the data distribution is same between shuffle and bucketed data source, which
       // enables us to only shuffle one side when join a bucketed table and a normal one.
-      HashPartitioning(bucketColumns, spec.numBuckets).partitionIdExpression
+      partitioning.partitionIdExpression
     }
+
+    // If the plan's outputPartitioning is the same as the the bucket spec, then each row will have
+    // a constant bucket id. We possibly can avoid the sort altogether.
+    val bucketSortExpression = plan.outputPartitioning match {
+      case output: HashPartitioning =>
+        bucketPartitioning.filterNot(_.semanticEquals(output)).map(_.partitionIdExpression)
```

Comment:
The advantage of the bucketing would be that on read you are going to maintain your bucketing information in the query planner, avoiding possibly very expensive shuffles if your downstream Spark jobs are aggregating or joining on the same key(s).

If I have just finished a join on the key in my original job, it doesn't make sense that I would need to re-sort the whole partition just to write it out.